libp2p/connmanager.nim

```diff
@@ -0,0 +1,276 @@
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import tables, sequtils, sets
+import chronos, chronicles, metrics
+import peerinfo,
+       stream/connection,
+       muxers/muxer
+
+declareGauge(libp2p_peers, "total connected peers")
+
+const MaxConnectionsPerPeer = 5
+
+type
+  TooManyConnections* = object of CatchableError
+
+  MuxerHolder = object
+    muxer: Muxer
+    handle: Future[void]
+
+  ConnManager* = ref object of RootObj
+    # NOTE: don't change to PeerInfo here
+    # the reference semantics on the PeerInfo
+    # object itself make it succeptible to
+    # copies and mangling by unrelated code.
+    conns: Table[PeerID, HashSet[Connection]]
+    muxed: Table[Connection, MuxerHolder]
+    cleanUpLock: Table[PeerInfo, AsyncLock]
+    maxConns: int
+
+proc newTooManyConnections(): ref TooManyConnections {.inline.} =
+  result = newException(TooManyConnections, "too many connections for peer")
+
+proc init*(C: type ConnManager,
+           maxConnsPerPeer: int = MaxConnectionsPerPeer): ConnManager =
+  C(maxConns: maxConnsPerPeer,
+    conns: initTable[PeerID, HashSet[Connection]](),
+    muxed: initTable[Connection, MuxerHolder]())
+
+proc contains*(c: ConnManager, conn: Connection): bool =
+  ## checks if a connection is being tracked by the
+  ## connection manager
+  ##
+
+  if isNil(conn):
+    return
+
+  if isNil(conn.peerInfo):
+    return
+
+  if conn.peerInfo.peerId notin c.conns:
+    return
+
+  return conn in c.conns[conn.peerInfo.peerId]
+
+proc contains*(c: ConnManager, peerId: PeerID): bool =
+  peerId in c.conns
+
+proc contains*(c: ConnManager, muxer: Muxer): bool =
+  ## checks if a muxer is being tracked by the connection
+  ## manager
+  ##
+
+  if isNil(muxer):
+    return
+
+  let conn = muxer.connection
+  if conn notin c:
+    return
+
+  if conn notin c.muxed:
+    return
+
+  return muxer == c.muxed[conn].muxer
+
+proc cleanupConn(c: ConnManager, conn: Connection) {.async.} =
+  ## clean connection's resources such as muxers and streams
+  ##
+
+  if isNil(conn):
+    return
+
+  if isNil(conn.peerInfo):
+    return
+
+  let peerInfo = conn.peerInfo
+  let lock = c.cleanUpLock.mgetOrPut(peerInfo, newAsyncLock())
+
+  try:
+    await lock.acquire()
+    trace "cleaning up connection for peer", peer = $peerInfo
+    if conn in c.muxed:
+      let muxerHolder = c.muxed[conn]
+      c.muxed.del(conn)
+
+      await muxerHolder.muxer.close()
+      if not(isNil(muxerHolder.handle)):
+        await muxerHolder.handle
+
+    if peerInfo.peerId in c.conns:
+      c.conns[peerInfo.peerId].excl(conn)
+
+      if c.conns[peerInfo.peerId].len == 0:
+        c.conns.del(peerInfo.peerId)
+
+    if not(conn.peerInfo.isClosed()):
+      conn.peerInfo.close()
+
+  finally:
+    await conn.close()
+    libp2p_peers.set(c.conns.len.int64)
+
+    if lock.locked():
```

Comment:
if `close()` we never get unlocked here, either lock goes before or completely outside `try/catch` I think for safety

Reply:
also I would use `withLock` helper, if we didn't remove it, as its interchangeable with nim std and anyway keeps track of lock unlock

Reply:
close is guaranteed not to throw

Reply:
> if close() fails we never get unlocked here, either lock goes before or completely outside try/catch I think for safety

close won't throw, unless it's a defect.

> also I would use withLock helper, if we didn't remove it, as its interchangeable with nim std and anyway keeps track of lock unlock

I thought we discarded `withLock` precisely because it conflicts with std?

Reply:
I think we never resolved that conversation. 
What's the conflict tho? Nim has a type system and allows overrides. I think just the auto had to be made an explicit type.

Reply:
> close is guaranteed not to throw

Can we document this somehow, I always wonder about those things, wish we had a nothrow attribute

Reply:
> What's the conflict tho? Nim has a type system and allows overrides. I think just the auto had to be made an explicit type.

Well, you're overriding a built in with an alternative that does something similar but in a completely different context, confusing at the very least. 

> Can we document this somehow, I always wonder about those things, wish we had a nothrow attribute

Well, that would be our `raises`, but it's a no-go for now, so I'm yet to come up with a succinct way of communicating it without inserting a `# does not throw` at each call site, yaiks!